runtime/caml/platform.h

```diff
@@ -100,22 +100,49 @@ Caml_inline uintnat atomic_fetch_add_verify_ge0(atomic_uintnat* p, uintnat v) {
   return result;
 }
 
+/* Warning: blocking functions.
+
+   Blocking functions are for use in the runtime outside of the
+   mutator, or when the domain lock is not held.
+
+   In order to use them inside the mutator and while holding the
+   domain lock, one must make sure that the wait is very short, and
+   that no deadlock can arise from the interaction with the domain
+   locks and the stop-the-world sections.
+
+   In particular one must not call [caml_plat_lock] on a mutex while
+   the domain lock is held:
+    - if any critical section of the mutex crosses an allocation, a
+      blocking section releasing the domain lock, or any other
+      potential STW section, nor
+    - if the same lock is acquired at any point using [Mutex.lock] or
+      [caml_plat_lock_non_blocking] on the same domain (circular
+      deadlock with the domain lock).
+
+   Thus, as a general rule, prefer [caml_plat_lock_non_blocking] to
+   lock a mutex when inside the mutator and holding the domain lock.
+   The domain lock must be held in order to call
+   [caml_plat_lock_non_blocking].
+
+   These functions never raise exceptions; errors are fatal. Thus, for
+   usages where bugs are susceptible to be introduced by users, the
+   functions from caml/sync.h should be used instead.
+*/
```

Comment:
It might make more sense, I think, to write this in terms of `caml_enter_blocking_section` and `caml_leave_blocking_section`, both since it describes what it does more precisely ("I can block within this section and it's OK"), and because the domain lock is sort of an implementation detail, which is less documented ("what is the domain lock? what does it protect?").

My understanding is that the issue is blocking without any C thread to handle possible interrupts (neither the main nor backup thread). It seems the domain lock must be held for handling interrupts (I suppose to ensure that the main thread cannot enter re-enter the mutator while the backup thread is handling interrupts, and that the backup thread doesn't handle interrupts before the mutator has been left). However it is not specifically the holding of the lock that prevents the backup thread from handling interrupts, since the backup thread must also be properly be in the `BT_IN_BLOCKING_SECTION` state.

Reply:
You correctly point out a subtlety about domain locks. In OCaml 4, the runtime lock was essentially a mutex, but in OCaml 5 the domain lock is an abstraction that also deals with waking up the backup thread.
That being said, a clarification about this would be better located in the manual about the domain lock (https://ocaml.org/manual/5.1/intfc.html#ss:parallel-execution-long-running-c-code).
I think this documentation is fine here because *domain lock*, `caml_release/acquire_runtime_system`, and `caml_enter_blocking_section/caml_enter_blocking_section` are described as corresponding to the same notion in the manual.

Reply:
Personally I find the terminology "{enter,leave} a blocking section" confusing, because "blocking section" may be interpreted as meaning two opposite things, "a section that blocks the runtime", or "a section that contains a blocking operation while making sure that the runtime is not blocked". The notion of taking/releasing the {runtime, domain} lock are slightly more low-level but also clearer in what they mean, so I find them easier to reason about.

My mental model:
- mutator threads must "own a domain" to use runtime services
- taking ownership is realized by taking the domain lock
- the owner of the domain is also responsible for handling STW interrupts, so it must ensure that it remains available to do so

In programs that run several threads on a single domain, releasing the domain lock lets other threads run and handle interrupts. In any case, the runtime ensures that there is always at least one other thread that is available (the backup thread).